flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -0,0 +1,347 @@
+        comment = SqlLiteral.createCharString(p, getPos());
+    }]
+    [
+        <PARTITIONED> <BY>
+            {
+                SqlNode column;
+                List<SqlNode> partitionKey = new ArrayList<SqlNode>();
+                pos = getPos();
+
+            }
+            <LPAREN>
+            [
+                column = SimpleIdentifier()
+                {
+                    partitionKey.add(column);
+                }
+                (
+                    <COMMA> column = SimpleIdentifier()
+                        {
+                            partitionKey.add(column);
+                        }
+                )*
+            ]
+            <RPAREN>
+            {
+                partitionColumns = new SqlNodeList(partitionKey, pos.plus(getPos()));
+            }
+    ]
+    [
+        <WITH>
+            {
+                SqlNode property;
+                List<SqlNode> proList = new ArrayList<SqlNode>();
+                pos = getPos();
+            }
+            <LPAREN>
+            [
+                property = PropertyValue()
+                {
+                proList.add(property);
+                }
+                (
+                <COMMA> property = PropertyValue()
+                    {
+                    proList.add(property);
+                    }
+                )*
+            ]
+            <RPAREN>
+        {  propertyList = new SqlNodeList(proList, pos.plus(getPos())); }
+    ]
+
+    {
+        return new SqlCreateTable(startPos.plus(getPos()),
+                tableType,
+                tableName,
+                columnList,
+                primaryKeyList,
+                uniqueKeysList,
+                watermark,
+                propertyList,
+                partitionColumns,
+                comment);
+    }
+}
+
+SqlNode SqlDropTable() :
+{
+    SqlParserPos pos;
+    SqlIdentifier tableName = null;
+    boolean ifExists = false;
+}
+{
+    <DROP> { pos = getPos(); }
+
+    <TABLE>
+
+    [<IF> <EXISTS> { ifExists = true; } ]
+
+    tableName = CompoundIdentifier()
+
+    {
+         return new SqlDropTable(pos, tableName, ifExists);
+    }
+}
+
+SqlIdentifier SqlArrayType() :
+{
+    SqlParserPos pos;
+    SqlDataTypeSpec elementType;
+}
+{
+    <ARRAY> { pos = getPos(); }
+    <LT> elementType = DataType()
+    <GT>
+    {
+        return new SqlArrayType(pos, elementType);
+    }
+}
+
+SqlIdentifier SqlMapType() :
+{
+    SqlParserPos pos;
+    SqlDataTypeSpec keyType;
+    SqlDataTypeSpec valType;
+}
+{
+    <MAP> { pos = getPos(); }
+    <LT> keyType = DataType()
+    <COMMA> valType = DataType()
+    <GT>
+    {
+        return new SqlMapType(pos, keyType, valType);
+    }
+}
+
+SqlIdentifier SqlStructType() :
+{
+    SqlParserPos pos;
+    List<SqlIdentifier> fieldNames = new ArrayList<SqlIdentifier>();
+    List<SqlDataTypeSpec> fieldTypes = new ArrayList<SqlDataTypeSpec>();
+}
+{
+    <STRUCT> { pos = getPos(); SqlIdentifier fName; SqlDataTypeSpec fType;}
```

Comment:
Please align the parser with the new type system described in FLIP-37: https://docs.google.com/document/d/1a9HUb6OaBIoj9IRfbILcMFPrOL7ALeZ3rVI66dvA2_U/edit#

The parser should be able to parse all serialized representations of the data types listed in `org.apache.flink.table.types.LogicalTypesTest`.

Data types such as `STRUCT` are not SQL standard. Instead, Flink builds on top of `ROW` and adds comment functionality there.

Reply:
Speaking standard, I'm wondering what's our stance on this, because in the same PR, I can see the create table statement, which is also way off standard. I was actually expecting a generic approach like putting the watermark, computed column info in the properties so as to keep the grammar closer to the standard. However, I'd like to share others' opinions.

I asked this because we also planned to support HIve's create table syntax, which has a lot of nonstandard flavors.

Reply:
@twalthr Ok, let's use `Row` as struct keyword